contrib/zenith/file_cache.c

```diff
@@ -0,0 +1,345 @@
+/*
+ *
+ * file_cache.c
+ *
+ *
+ * Portions Copyright (c) 1996-2021, PostgreSQL Global Development Group
+ * Portions Copyright (c) 1994, Regents of the University of California
+ *
+ *
+ * IDENTIFICATION
+ *	  contrib/zenith/file_cache.c
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include <sys/file.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "postgres.h"
+#include "miscadmin.h"
+#include "pgstat.h"
+#include "pagestore_client.h"
+#include "postmaster/bgworker.h"
+#include "storage/relfilenode.h"
+#include "storage/buf_internals.h"
+#include "storage/latch.h"
+#include "storage/ipc.h"
+#include "storage/lwlock.h"
+#include "utils/dynahash.h"
+#include "utils/guc.h"
+#include "storage/fd.h"
+#include <storage/buf_internals.h>
+
+/*
+ * Local file cache is used to temporary store relations pages in local file system.
+ * All blocks of all relations are stored inside one file and addressed using shared hash map.
+ * Currently LRU eviction policy based on L2 list is used as replacement algorithm.
+ * As far as manipulation of L2-list requires global critical section, we are not using partitioned hash.
+ * Also we are using exclusive lock even for read operation because LRU requires relinking element in L2 list.
+ * If this lock become a bottleneck, we can consider other eviction strategies, for example clock algorithm.
+ *
+ * Cache is always reconstructed at node startup, so we do not need to save mapping somewhere and worry about
+ * its consistency.
+ */
+
+/* Local file storage allocation chunk.
+ * Should be power of two. Using larger than page chunks can
+ * 1. Reduce hash-map memory footprint: 8TB database contains billion pages
+ *    and size of hash entry is 40 bytes, so we need 40Gb just for hash map.
+ *    1Mb chunks can reduce hash map size to 320Mb.
+ * 2. Improve access locality, subsequent pages will be allocated together improving seqscan speed
+ */
+#define CHUNK_SIZE 128 /* 1Mb chunk */
+
+typedef struct FileCacheEntry
+{
+	BufferTag key;
+	uint32    offset;
+	uint32    access_count;
+	uint32    bitmap[CHUNK_SIZE/32];
```

Comment:
You should probably do some static asserts that `CHUNK_SIZE >= 32`, as a CHUNK_SIZE <32 will result in a 0-length array here.

Reply:
By static assert you mean:
```
#if  CHUNK_SIZE < 32 
   #error Invalid CHUNK_SIZE
#endif
```
IMHO it is overkill, taken in account that it is defined few lines upper. I will mention in comment that it should not be less than 32.

Reply:
With static assert I was refering to the `StaticAssertStmt` / `StaticAssertDecl` as defined in postgres/src/include/c.h, or the native static assertion builtins provided by compilers.